### ISSUE TITLE
All criteria operators are converted to = when queried via the ORM

### ISSUE THREAD

**Issue by jaikdean** [doctrine/doctrine2#5687]
**Version:** 2.5.4

**Example**

``` php
$criteria = Criteria::create()->where(
    Criteria::expr()->neq('id', $idToFilter)
);

return $entity->getAssociatedEntities()->matching($criteria);
```

**Expected result**
All associated entities except for ones with an `id` of `$idToFilter` will be returned.

**Actual result**
Only entities with an `id` of `$idToFilter` will be returned. In fact any expression, such as `neq` or `notIn` will end up making an `=` comparison in the SQL query.

**Investigation**

I've walked through the code and tracked down the probably culprit to `Doctrine\ORM\Persisters\Collection\ManyToManyPersister::loadCriteria()`, lines 241–245:

``` php
        foreach ($parameters as $parameter) {
            list($name, $value) = $parameter;
            $whereClauses[]     = sprintf('te.%s = ?', $name);
            $params[]           = $value;
        }
```

This always uses the `=` operator, regardless of the one specified.

`Doctrine\ORM\Persisters\SqlValueVisitor::walkComparison()` gets the operator as a local variable which it uses to alter behaviour if we're doing a comparison with `null`, but then discards it. I believe this is the earliest point in the chain that the operator is effectively lost.

**Comment by TheFox** [comment]
I have the same issue. I thought maybe I can use the following workaround for `id <> 21` expression: `id < 21 or id > 21`. But this is also not working because the `AND` is hard-coded in `loadCriteria()`: `' WHERE ' . implode(' AND ', $whereClauses)`.

**Comment by flip111** [comment]
It would help to know which class `$entity->getAssociatedEntities()` is. I would assume a collection of the orm and not a more general type of collection such as [ArrayCollection](https://github.com/doctrine/collections/blob/2dfc898f2a66c192d6a195bdc9c5040f65adfaab/lib/Doctrine/Common/Collections/ArrayCollection.php). The `ManyToManyPersister` class you mention, and the `loadCriteria` function in particular cause [queries to the database](https://github.com/doctrine/doctrine2/blob/cd11723e6310b5973f50e7d9819c58916721d290/lib/Doctrine/ORM/Persisters/Collection/ManyToManyPersister.php#L274-L284). So it seems an orm collection would eventually trigger a call to `ManyToManyPersister->loadCriteria`.

The criteria class never really supposed to work in this way. This is a known but undocumented quirk. I first saw it about 3 years ago. At least if my memory serves me correctly ...

The reason behind it is that it will break some of the ORM stuff when the collection is incomplete. That is .. a partial collection (compared to the "full" database table). My best guess is this is because there are some methods who suppose to work on the entire collection (what else could it be really?).

I don't think it's technically impossible to come with a solution here, but if you want to work with a partial collection and some functions need the full collection you would have to keep reference to two collections for the same associated entity. And in case you query the partial collection first, you might have to re-query the entire collection later.

In SQL partial collections would work out in a weird way as well. Suppose you have 1 entity "tab1" owning a collection of "tab2", you would query `SELECT tab1.*, tab2.* FROM tab1 JOIN tab2 ON tab1.id = tab2.tab1_id where tab1.id = 1`, fine. Now ask doctrine please load me all the associated "tab3" you can find as child of "tab2", ok another query `SELECT tab2.*, tab3.* FROM tab2 JOIN tab3 ON tab2.id = tab2.tab3_id`. But what if you first loaded "tab2" with a criteria, doctrine would have to store this clause, and use it later (possibly as sub-query). Then if you chain this a couple of times (tab4, tab5, tab6 ...) the queries you get would get pretty ridiculous.

As workaround you could possibly convert your collection to the more general ArrayCollection. At least don't use your ORM collection for anything other than READING from the database. If you don't care to use criteria at all you could just as well get the results as an array which should be a bit faster as well.

**Comment by jaikdean** [comment]
I think this has been fixed by #6740
